### PR TITLE
User config watcher

### PR DESCRIPTION
**Pull request by asymmetric**
This PR adds a watcher for the `user.toml` file.

It's based on @krnowak's still-unmerged work on the file watcher.

### CODE REVIEW

File: components/sup/src/manager/service/mod.rs [hunk 1]
```diff
@@ -660,6 +662,7 @@ impl Service {
     }
 
     /// Write service files from gossip data to disk.
+    /// For the location, check `svc_file_path()`.
```
Comment (asymmetric) [inline components/sup/src/manager/service/mod.rs]: Wasn't sure how to create a symlink for this. Also, the function is private, so I'm not sure if there's a point.

**Comment by asymmetric** [comment]
**NB**: The config is not currently reloaded, so this doesn't work yet.

File: components/sup/src/manager/mod.rs [hunk 2]
```diff
@@ -277,12 +280,13 @@ impl Manager {
...
             launcher: launcher,
             services: services,
-            watcher: SpecWatcher::run(&fs_cfg.specs_path)?,
+            spec_watcher: SpecWatcher::run(&fs_cfg.specs_path)?,
             fs_cfg: Arc::new(fs_cfg),
             organization: cfg.organization,
             service_states: HashMap::new(),
             sys: Arc::new(sys),
             peer_watcher: peer_watcher,
+            user_config_watcher: UserConfigWatcher::new(),
```
Comment (krnowak) [inline components/sup/src/manager/mod.rs]: If you changed the order of members in `Manager` struct, it would be good to reorder the initializations here too.

File: components/sup/src/manager/service/mod.rs (hunk 1 above)
Comment (krnowak) [reply components/sup/src/manager/service/mod.rs]: I guess this is a note to yourself, because I don't understand what it is about. Not enough context.

File: components/sup/src/manager/user_config_watcher.rs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    }
+
+    fn service_group(&self) -> &ServiceGroup {
+        &self.service_group
+    }
+}
+
+
+pub struct UserConfigWatcher {
+    // TODO use a HashSet? we don't really need the boolean:
```
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: This comment does not seem to be true - `reset_events_for` sets the boolean to false.

File: components/sup/src/manager/user_config_watcher.rs [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        ThreadBuilder::new()
+            .name(format!("user-config-watcher-{}", path.display()))
+            .spawn(move || {
+                debug!(
+                    "UserConfigWatcher({}) worker thread starting",
+                    path.display()
+                );
+                let callbacks = UserConfigCallbacks { have_events: have_events };
+                let mut file_watcher =
+                    default_file_watcher(&path, callbacks).expect("creating user-config watcher");
```
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: Instead of using `expect`, I'd use `outputln!` macro to report an error and maybe quit the thread. I'm not sure if we should panic in this case.

File: components/sup/src/manager/user_config_watcher.rs [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    fn events_present_after_removing_config() {
+        let service = TestService::default();
+        let file_path = service.path().join(USER_CONFIG_FILE);
+        let mut ucm = UserConfigWatcher::new();
+
+        ucm.add(&service);
+        File::create(&file_path).expect("creating file");
+
+        // Allow the watcher to notice that a file was created.
+        thread::sleep(Duration::from_millis(100));
```
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: Why sleeping instead of `wait_for_events`?

File: components/sup/src/manager/user_config_watcher.rs [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::thread::Builder as ThreadBuilder;
+
+use super::file_watcher::{Callbacks, default_file_watcher};
+
+use hcore::fs::USER_CONFIG_FILE;
+use hcore::service::ServiceGroup;
+use manager::service::Service;
+
+static LOGKEY: &'static str = "UCM";
```
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: Consider adding this log key to the documentation here - `www/source/docs/reference/log-keys.html.md`.

**Review by krnowak — CHANGES_REQUESTED** [review]
Looks good overall, I had some nits.

File: components/sup/src/manager/service/mod.rs (hunk 1 above)
Comment (asymmetric) [reply components/sup/src/manager/service/mod.rs]: The `svc_file_path` function could be rendered as a link, since `///` comments are Markdown. I just don't know what the target should be, as it seems to be relative to something.

File: components/sup/src/manager/user_config_watcher.rs (hunk 3 above)
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: Yes, this is a note to myself saying that we could change the implementation. True/false values can be represented by presence/absence in a set.

File: components/sup/src/manager/user_config_watcher.rs [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        }
+    }
+
+    impl Default for TestService {
+        fn default() -> Self {
+            Self {
+                name: String::from("foo"),
+                path: TempDir::new("user-config-watcher")
+                    .expect("creating temp dir")
+                    .into_path(),
```
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: One note about this - when you use the `into_path()` function, you also take over the responsibility of removing the created temporary directory. Which you probably don't do, no?

File: components/sup/src/manager/user_config_watcher.rs [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+use hcore::fs::USER_CONFIG_FILE;
+use hcore::service::ServiceGroup;
+use manager::service::Service;
+
+static LOGKEY: &'static str = "UCM";
+
+// This trait exists to ease the testing of functions that receive a Service.  Creating Services
+// requires a lot of ceremony, so we work around this with this trait.
+pub trait Serviceable {
+    fn name(&self) -> &String;
```
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: Returning references makes the trait less flexible, because you need an existing, non-temporary thing in memory to return its address (like a member of a struct that implements this trait). In other words - you can't compute the return value. Not sure if it is really important, but in general I would just return plain values. Of course the downside of returning values is that you need to clone the stuff, but OTOH this won't likely be a bottleneck ever.

**Review by krnowak — CHANGES_REQUESTED** [review]
Another two things I noticed.

File: components/sup/src/manager/user_config_watcher.rs (hunk 8 above)
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: My rationale here is that the name is always a reference to its containing struct.

Isn't it enough for users of this trait to be able to call `name.clone()` if they want to own the `String`?

Can you show an example of where the current implementation creates issues? As I'm not sure I understand your comment.

File: components/sup/src/manager/user_config_watcher.rs (hunk 4 above)
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: Good point, I've committed a change.

File: components/sup/src/manager/user_config_watcher.rs (hunk 7 above)
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: Good point, fixed.

File: components/sup/src/manager/user_config_watcher.rs (hunk 8 above)
Comment (krnowak) [reply components/sup/src/manager/user_config_watcher.rs]: There are no problems with current implementation. It was just a general note about returning references in trait functions. So in this case it's likely ok, but in general I would avoid it.

File: components/sup/src/manager/service/mod.rs (hunk 1 above)
Comment (asymmetric) [reply components/sup/src/manager/service/mod.rs]: Fixed this.

File: components/sup/src/manager/user_config_watcher.rs (hunk 3 above)
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: Tried using a `HashSet` and turns out it's more complicated than expected (if we put a `HashSet` in an `Arc`, that `HashSet` is immutable, unless I use mutexes), so not worth it.

File: components/sup/src/manager/user_config_watcher.rs (hunk 8 above)
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: OK, let me know if you feel strongly about changing this, otherwise I'm gonna leave it as is.

**Comment by asymmetric** [comment]
@krnowak addressed your comments, but I guess this will just wait until your PR gets merged.

I also rebased on yours.

I will clean up the commit messages next week.

File: components/sup/src/manager/user_config_watcher.rs (hunk 3 above)
Comment (krnowak) [reply components/sup/src/manager/user_config_watcher.rs]: Yeah, you would need to do something similar to what some code in butterfly does - `Arc<RwLock<HashSet<Whatever>>>`. Anyway, I think that current solution is fine as well.

File: components/sup/src/manager/user_config_watcher.rs (hunk 8 above)
Comment (krnowak) [reply components/sup/src/manager/user_config_watcher.rs]: Nah, leave it as is, please. If somebody else points out this thing, then you can change it.

File: components/sup/src/manager/service/mod.rs [hunk 9]
```diff
@@ -387,29 +392,33 @@ impl Service {
...
-        let cfg_updated = self.cfg.update(census_group);
+        let cfg_updated = self.cfg.update(census_group) || self.user_config_updated;
         if cfg_updated || census_ring.changed {
+            if let Err(e) = self.cfg.load_user(&self.pkg) {
+                outputln!(preamble self.service_group, "Loading user-config failed: {}", e);
+            }
             let (reload, reconfigure) = {
                 let ctx = self.render_context(census_ring);
-                let reload = self.compile_hooks(&ctx);
+                let reload = self.compile_hooks(&ctx) || self.user_config_updated;
```
Comment (krnowak) [inline components/sup/src/manager/service/mod.rs]: Should this `|| self.user_config_updated` be moved one line below, to the `let reconfigure` line?

File: components/sup/src/manager/user_config_watcher.rs [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+pub struct UserConfigWatcher {
+    states: HashMap<String, Arc<AtomicBool>>,
+}
+
+impl UserConfigWatcher {
+    pub fn new() -> Self {
+        Self { states: HashMap::new() }
+    }
+
+    pub fn add<T: Serviceable>(&mut self, service: &T) {
```
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: We have an `add` function. Do we need a `remove` function then?

**Review by krnowak — COMMENTED** [review]
I had two questions, otherwise looks good.

**Comment by asymmetric** [comment]
The PR is in a weird state right now. It shows commit that are not in the PR's branch. What happened is that the base branch got merged into master, and after that happened, I rebased my branch on master.

I then changed the PR to use master as the base branch, but now it shows commits here that shouldn't be in this branch (and aren't there according to `git log`).
It seems like this is a problem with GitHub's UI.

Also, please don't be too picky about commit messages (and the commits themselves), as I am still testing this PR out.

File: components/sup/src/manager/service/mod.rs (hunk 9 above)
Comment (asymmetric) [reply components/sup/src/manager/service/mod.rs]: No, the `needs_reconfiguration` flag is only used to [trigger a `reconfigure` hook](https://github.com/habitat-sh/habitat/blob/c827ae61fa9077974e22fdbf04aecbafefbcef19/components/sup/src/manager/service/mod.rs#L639-L646), which may not be defined. The `reload` flag is actually going to trigger a service reload, which is what we need here.

**Comment by asymmetric** [comment]
@krnowak could you take a look again?

File: components/sup/src/manager/mod.rs [hunk 11]
```diff
@@ -879,20 +904,30 @@ impl Manager {
...
         self.butterfly.set_departed();
 
-        let mut services = self.services.write().expect("Services lock is poisend!");
+        let mut svcs = Vec::new();
 
-        for mut service in services.drain(..) {
+        // The problem we're trying to work around here by adding this block is that `write`
+        // creates an immutable borrow on `self`, and `self.remove_service` needs `&mut self`.
+        // The solution is to introduce the block to drop the borrow before the call to
+        // `self.remove_service`, and use `mem::swap` to copy the services to a variable defined
```
Comment (krnowak) [inline components/sup/src/manager/mod.rs]: To move, not copy. After swapping, the vector under `self.services` lock will be empty.

File: components/sup/src/manager/service/mod.rs [hunk 12]
```diff
@@ -404,29 +409,49 @@ impl Service {
...
         let census_group = census_ring.census_group_for(&self.service_group).expect(
             "Service update failed; unable to find own service group",
         );
-        let cfg_updated = self.cfg.update(census_group);
-        if cfg_updated || census_ring.changed() {
+        let cfg_updated_from_rumors = self.cfg.update(census_group);
+        let cfg_changed = cfg_updated_from_rumors || self.user_config_updated;
+        let something_to_do = cfg_changed || census_ring.changed();
+
+        if something_to_do {
```
Comment (krnowak) [inline components/sup/src/manager/service/mod.rs]: This became a bit hard to follow. Would pseudo-rust below be clearer? (`cufr` = `cfg_updated_from_rumors`, `sucu` = `self.user_config_updated`, `crc` = `census_ring.changed()`)

```
if sucu {
    if let Err(e) = self.cfg.load_user(&self.pkg) {…}
}
let (cfg_changed, reload, reconfigure) = {
    let ctx = …;
    if cufr || crc {
        (cufr || sucu, self.compile_hooks(&ctx) || sucu, self.compile_configuration(&ctx))
    } else if sucu {
        (true, true, self.compile_configuration(&ctx))
    } else {
        (false, self.needs_reload, self.needs_reconfiguration)
    }
}
self.needs_reload = reload;
self.needs_reconfiguration = reconfigure;
self.user_config_updated = false;
/* return */ cfg_changed
```

File: components/sup/src/manager/user_config_watcher.rs [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+    // The channel is a SyncChannel with buffer size 1, as we are only interested in the fact that there were events,
+    // not how many there were.
+    have_events: Receiver<()>,
+    // This sender is used by the watcher to notify a worker to stop running.
+    // It is an async channel because we never want the UserConfigWatcher to block, even if the receiver end of
+    // the channel somehow dies and/or fails to consume the message.
+    stop_running: Sender<()>,
+}
+
+type ServiceName = String;
```
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs:64]: That's nice.

File: components/sup/src/manager/user_config_watcher.rs [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+                stop_running: running_tx,
+            };
+
+            self.states.insert(service.name().to_owned(), state);
+        }
+
+        Ok(())
+    }
+
+    /// Removes a service from the User Config Watcher, thereby stopping the watcher thread.
```
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: I'd add "eventually" - "thereby eventually stopping …"

File: components/sup/src/manager/user_config_watcher.rs [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+    pub fn remove<T: Serviceable>(&mut self, service: &T) -> Result<(), SendError<()>> {
+        if let Some(state) = self.states.remove(service.name()) {
+            state.stop_running.send(())?;
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether the watcher for the specified service has observed any events, thereby
```
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: thereby is your new favourite word, eh? :)

File: components/sup/src/manager/user_config_watcher.rs [hunk 16]
```diff
@@ -0,0 +1,338 @@
...
+    /// consuming them.
+    pub fn have_events_for<T: Serviceable>(&self, service: &T) -> bool {
+        if let Some(state) = self.states.get(service.name()) {
+
```
Comment (krnowak) [inline components/sup/src/manager/user_config_watcher.rs]: Drop this empty line?

**Review by krnowak — CHANGES_REQUESTED** [review]
Looks mostly fine, but there was one hairy piece of code I'd like to see if it can be "balder".

File: components/sup/src/manager/service/mod.rs (hunk 12 above)
Comment (krnowak) [reply components/sup/src/manager/service/mod.rs]: The `something_to_do` variable would be gone.

Comment (asymmetric) [reply components/sup/src/manager/service/mod.rs]: I find the 3 value tuple even less clear tbh...

> (cufr || sucu, self.compile_hooks(&ctx) || sucu, self.compile_configuration(&ctx))

what?

File: components/sup/src/manager/user_config_watcher.rs (hunk 15 above)
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: It's a pretty cool word, you gotta admit that

File: components/sup/src/manager/user_config_watcher.rs (hunk 14 above)
Comment (asymmetric) [reply components/sup/src/manager/user_config_watcher.rs]: Why eventually? Do you mean that the operation is asynchronous?

If so, I think I can explain it more explicitly.

File: components/sup/src/manager/service/mod.rs (hunk 12 above)
Comment (krnowak) [reply components/sup/src/manager/service/mod.rs]: Not sure what this "what?" is about, but to be clear - I don't actually want to have variables like `cufr` or `sucu`. I wrote it like that to avoid typing in the comment.

Comment (asymmetric) [reply components/sup/src/manager/service/mod.rs]: I think that even with more descriptive names, it's hard to parse logically. There's just too much happening on one line: returning a tuple, ORing two values, calling functions..

Not sure it's an improvement.

Comment (krnowak) [reply components/sup/src/manager/service/mod.rs]: You don't need to put everything in one line. You could very well have:
```
let config_changed = config_updated_from_rumors || self.user_config_updated;
let needs_reload = self.compile_hooks(&ctx) || self.user_config_updated;
let needs_reconfigure = self.compile_configuration(&ctx);
(config_changed, needs_reload, needs_reconfigure)
```

In general, what I wanted to propose is to simplify the flow of the code, instead of using nested `if`s and `match`es that are sometimes executed and sometimes not.

Comment (asymmetric) [reply components/sup/src/manager/service/mod.rs]: In the above implementation though, we're calling `self.compile_hooks` even if we don't need to, which will lead to "Hooks recompiled" being printed over and over (this function gets executed [on every tick](https://github.com/kinvolk/habitat/blob/0a1d210328174ae822c13213e2462f6f37f1e94c/components/sup/src/manager/service/mod.rs#L294)).

I'm definitely open to simplifying the code but obviously it has to have the same behaviour.

The problem is that some of these functions have side effects so we can't just call them willy-nilly.

Comment (krnowak) [reply components/sup/src/manager/service/mod.rs]: The `self.compile_hooks(&ctx)` should only be executed when either `cfg_updated_from_rumors` is true or if `census_ring.changed()` returns true. So if this happens everytime, then something is wrong.

**Comment by asymmetric** [comment]
Closing in favor of https://github.com/habitat-sh/habitat/pull/3898.
